pageserver/src/tenant/delete.rs

```diff
@@ -0,0 +1,574 @@
+                init_order,
+                // Important. We dont pass ancestor above because it can be missing.
+                // Thus we need to skip the validation here.
+                false,
+            )
+            .context("create_timeline_struct")?;
+
+        let mut guard = DeletionGuard(
+            Arc::clone(&timeline.delete_progress)
+                .try_lock_owned()
+                .expect("cannot happen because we're the only owner"),
+        );
+
+        // Note: here we even skip populating layer map. Timeline is essentially uninitialized.
+        // RemoteTimelineClient is the only functioning part.
+        timeline.set_state(TimelineState::Stopping);
+        // We meed to do this because when console retries delete request we shouldnt answer with 404
+        // because 404 means successful deletion.
+        // FIXME consider TimelineState::Deleting.
+        {
+            let mut locked = tenant.timelines.lock().unwrap();
+            locked.insert(timeline_id, Arc::clone(&timeline));
+        }
+
+        guard.mark_in_progress()?;
+
+        Self::schedule_background(guard, tenant.conf, tenant, timeline);
+
+        Ok(())
+    }
+
+    pub async fn cleanup_remaining_fs_traces_after_timeline_deletion(
+        tenant: &Tenant,
+        timeline_id: TimelineId,
+    ) -> anyhow::Result<()> {
+        cleanup_remaining_fs_traces_after_timeline_deletion(
+            tenant.conf,
+            tenant.tenant_id,
+            timeline_id,
+        )
+        .await
+    }
+
+    fn prepare(
+        tenant: &Tenant,
+        timeline_id: TimelineId,
+    ) -> Result<(Arc<Timeline>, DeletionGuard), DeleteTimelineError> {
+        let timelines = tenant.timelines.lock().unwrap();
+
+        let timeline = match timelines.get(&timeline_id) {
+            Some(t) => t,
+            None => return Err(DeleteTimelineError::NotFound),
+        };
+
+        // Ensure that there are no child timelines **attached to that pageserver**,
+        // because detach removes files, which will break child branches
+        let children: Vec<TimelineId> = timelines
+            .iter()
+            .filter_map(|(id, entry)| {
+                if entry.get_ancestor_timeline_id() == Some(timeline_id) {
+                    Some(*id)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if !children.is_empty() {
+            return Err(DeleteTimelineError::HasChildren(children));
+        }
+
+        let delete_lock_guard = DeletionGuard(
+            Arc::clone(&timeline.delete_progress)
+                .try_lock_owned()
+                .map_err(|_| {
+                    DeleteTimelineError::Other(anyhow::anyhow!(
+                        "timeline deletion is already in progress"
+                    ))
+                })?,
+        );
+
+        // FIXME is it really possible given that we remove timeline under mutex when we complete the deletion and
```

Comment:
This place needs attention

Reply:
Realized that with my rebase of original "delete from s3 PR" I accidentally dropped the line that set deletion lock value to true.

I believe that this situation should be impossible, because we lock the guard while we hold timelines lock, so if we started the deletion you cant get the timeline form timelines map and see the deletion guard unlocked if deletion is still ongoing.

The original comment says that it can happen if we just finished the deletion, but if we remove timeline from timelines map fitst and then drop the guard this shouldnt be possible. Guard is either locked and deletion is ongoing, or timeline is removed from timelines map and cannot be retrieved at this point.

Do you agree with this logic? @koivunej 

cc @hlinnaka

Reply:
> I believe that this situation should be impossible, because we lock the guard while we hold timelines lock

But we don't hold the timelines lock over the whole dance now do we? Only during this prepare phase. And after that, it's the request coalescing job, which is the `timeline.delete_progress` but ... we don't actually stick around to wait for that either, since there's the try_lock_owned and AlreadyInProgress error returned quick.

Reply:
> But we don't hold the timelines lock over the whole dance now do we?

Which dance you're referring to? The whole deletion process? As I see it there is a clear happens before relationship. Timeline is in the map -> check the guard, if locked return.  When we finish the deletion we remove timeline from the map first and then drop the guard, so I dont see how we can get a timeline with Finished guard from the map

Reply:
Thanks! Yeah, panics should be ok because we use tokio mutex for deletion guard, so no poisoning should happen